StreamChatClient_v3/APIClient/Endpoints/ChannelEndpoints.swift

```diff
@@ -0,0 +1,305 @@
+//
+// ChannelEndpoints.swift
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+extension Endpoint {
+    static func channels<ExtraData: ExtraDataTypes>(query: ChannelListQuery)
+        -> Endpoint<ChannelListEndpointPayload<ExtraData>> {
+        .init(path: "channels",
+              method: .get,
+              queryItems: [],
+              jsonQueryItems: ["payload": query],
+              body: nil)
+    }
+}
+
+struct ChannelListEndpointPayload<ExtraData: ExtraDataTypes>: Decodable {
+    /// A list of channels response (see `ChannelQuery`).
+    let channels: [ChannelEndpointPayload<ExtraData>]
+}
+
+struct ChannelEndpointPayload<ExtraData: ExtraDataTypes>: Decodable {
+    let channel: ChannelDetailPayload<ExtraData>
+    
+    let watcherCount: Int
+    
+    let members: [MemberEndpointPayload<ExtraData.User>]
+    
+    // TODO:
+    /*
+     messages
+     message reads
+     
+     */
+    private enum CodingKeys: String, CodingKey {
+        case channel
+//    case messages
+//    case messageReads = "read"
+        case members
+//    case watchers
+        case watcherCount = "watcher_count"
+    }
+    
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        channel = try container.decode(ChannelDetailPayload<ExtraData>.self, forKey: .channel)
+        watcherCount = try container.decode(Int.self, forKey: .watcherCount)
+        members = try container.decode([MemberEndpointPayload<ExtraData.User>].self, forKey: .members)
+    }
+    
+    // MARK: - For testing
+    
+    init(channel: ChannelDetailPayload<ExtraData>, watcherCount: Int, members: [MemberEndpointPayload<ExtraData.User>]) {
+        self.channel = channel
+        self.watcherCount = watcherCount
+        self.members = members
+    }
+}
+
+struct ChannelDetailPayload<ExtraData: ExtraDataTypes>: Decodable {
+    let cid: String
+    
+    let extraData: ExtraData.Channel?
+    
+    /// A channel type.
+    public let typeRawValue: String
+    /// A channel id.
+    public let id: String
```

Comment:
This is just the DTO object. It wouldn't make sense to serialize it here and then again deserialize it for storing in the db. `Payload` objects should match the incoming JSON as closely as possible.

Reply:
I think we should use the concept of `cid` everywhere. This is really important. `ChannelId` is a simple decodable type.
You can't do that:
```swift
let dto = ChannelDTO.loadOrCreate(id: payload.channel.id, context: self)
```
Because could be channels with the same id, like `general` for different types.

Reply:
Great! We talked about it in Slack a little bit already. I think it makes sense 👍

Reply:
Yes we talked about that and decided to use cid everywhere.